### DISCUSSION TITLE
[wasm-split] Push proxied calls to primary module

### DISCUSSION THREAD

**Discussion by Mintyboi** [emscripten-core/emscripten#21374] [General]
I'm building my application with threads and and with the `sPROXY_TO_PTHREAD` flag.

When we split the modules (`-sSPLIT_MODULE`), there are some functions that have to be present in the primary module, mainly the proxied functions like syscalls, webgpu and pthread (the one I'm most interested in) operations.

Is there a deterministic way to collect all these proxied functions and have them kept in the primary module?
I though of using a callgraph to find the functions that are needed to create a list, however there are too many workflows and this lists needs to be maintained.

cc @sbc100 @tlively

**Comment by sbc100** [comment]
I believe the two main options you have for decided where to split are call graphs analysis and profiling.   

The callgraph method is the likely the most reliable method but could end up pulling more than you would want into the main module.   I would hope that you would automate it and therefore the cost of maintenance would not be significant.

**Comment by tlively** [comment]
The most robust solution would be to add a new option to the `wasm-split` tool to have it do the callgraph analysis for you based on a list of entry points you provide. This could perhaps be paired with the profiling approach so you can get the best of both worlds.

cc @dschuff, with whom I was recently discussing options in this space.

**Comment by Mintyboi** [comment]
> new option to the wasm-split tool to have it do the callgraph analysis for you based on a list of entry points you provide

@tlively Yes, this is something useful that would help!
I have been tweaking wasm-opt to allow us to be able to filter the callgraph to only the entry points that we specified

https://github.com/WebAssembly/binaryen/pull/6354
